src/udp.c

```diff
@@ -133,6 +132,46 @@ static int udp_input(const struct ip_hdr *ip_hdr,
 }
 IP_PROTO_INPUT_HANDLER(IP_PROTO_UDP, udp_input);
 
+static uint16_t udp_csum(const void *buff,
+                  size_t len,
+                  in_addr_t src_addr,
+                  in_addr_t dest_addr)
+{
+    const uint16_t *buf = buff;
+    uint16_t *ip_src = (void *) &src_addr, *ip_dst = (void *) &dest_addr;
+    uint32_t sum;
+    size_t length = len;
+
+    // Calculate the sum
+    sum = 0;
+    while (len > 1) {
+        sum += *buf++;
+        if (sum & 0x80000000)
+            sum = (sum & 0xFFFF) + (sum >> 16);
+        len -= 2;
+    }
+    // Add the padding if the packet lenght is odd
+    if (len & 1)
+        sum += *((uint8_t *) buf);
+
+    // Add the pseudo-header
```

Comment:
The following lines look ugly at first glance. You should explain with more comments.